src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs

```diff
@@ -58,60 +62,139 @@ public class QueryService : IQueryService
-                var seriesResultMap = seriesComputedResults.ToDictionary<SeriesResult, string>(a => a.StudyInstanceUid + "-" + a.SeriesInstanceUid, StringComparer.OrdinalIgnoreCase);
-                responseMetadata = await Task.WhenAll(queryResult.DicomInstances.Select(x =>
-                GenerateQueryResult(x, seriesResultMap[x.StudyInstanceUid + "-" + x.SeriesInstanceUid].DicomDataset, responseBuilder, getFullMetadata: true, cancellationToken)));
+                getStudyResponse = false;
+                getSeriesResponse = false;
             }
         }
-        if (responseMetadata == null)
+
+        // logging to track usage
+        _logger.LogInformation("QueryService result retrieval resultCount:{ResultCount}, studyResultRetrieved:{StudyResultRetrieved}, seriesResultRetrieved:{SeriesResultRetrieved}, fullMetadataRetrieved:{FullMetadataRetrieved}", queryResult.DicomInstances.Count(), getStudyResponse, getSeriesResponse, getFullMetadata);
+
+        // start getting and merging the results based on the source.
+        IEnumerable<DicomDataset> instanceMetadata = null;
+        var versions = queryResult.DicomInstances.Select(i => i.Version).ToList();
+        if (getFullMetadata)
         {
-            IEnumerable<DicomDataset> instanceMetadata = await Task.WhenAll(
+            instanceMetadata = await Task.WhenAll(
                 queryResult.DicomInstances.Select(x => _metadataStore.GetInstanceMetadataAsync(x, cancellationToken)));
-
-            responseMetadata = instanceMetadata.Select(m => responseBuilder.GenerateResponseDataset(m));
         }
+        if (getSeriesResponse)
+        {
+            var seriesComputedResults = await _queryStore.GetSeriesResultAsync(partitionKey, versions, cancellationToken);
 
-        return new QueryResourceResponse(responseMetadata, queryExpression.ErroneousTags);
-    }
-
-    // TODO version the API to change the default results to only include indexed columns and control the getFullMetadata param
-    private async Task<DicomDataset> GenerateQueryResult(
-        VersionedInstanceIdentifier versionedInstance,
-        DicomDataset computedMetadata,
-        QueryResponseBuilder queryResponseBuilder,
-        bool getFullMetadata,
-        CancellationToken cancellationToken)
-    {
-        if (getFullMetadata)
+            if (instanceMetadata == null)
+            {
+                instanceMetadata = seriesComputedResults.Select(x => x.DicomDataset);
+            }
+            else
+            {
+                var map = seriesComputedResults.ToDictionary<SeriesResult, string>(a => a.StudyInstanceUid + "-" + a.SeriesInstanceUid, StringComparer.OrdinalIgnoreCase);
+                instanceMetadata = instanceMetadata.Select(x =>
+                {
+                    var ds = new DicomDataset(x);
+                    return ds.AddOrUpdate(map[x.GetSingleValue<string>(DicomTag.StudyInstanceUID) + "-" + x.GetSingleValue<string>(DicomTag.SeriesInstanceUID)].DicomDataset);
```

Comment:
I'm a little leery of using the `-` character as a separator here - currently we do not allow UIDs to have that character but if we do in the future as part of leniency this logic would break.

Reply:
the separator character does not matter here.

Reply:
Can we instead use my `DicomIdentifier` struct here instead? I implemented `GetHashCode` and `Equals` so it should be ok to use in a dictionary.